Add LeakyRelu C++ Op and its gradient implementation.

lowintelligence opened the pull request:
LeakyRelu, defined as 'y = { x (x>=0) or alpha*x (x<0) }', was computed by combined Ops 'max(x, alpha*x)' in current codes. Hence its gradient calculation for back propagation would contain a serial of element-wise Ops. This looks really unnecessary for such a simple op and it could be done within just one Op with less memory accesses.

In fact, we meet some performance issue on CPU in an attention based model. Within the FC calculation 'Leaky ReLu' was used as the activation function, so that its back prop Ops took some CPU time. Thus we decided to implement leaky_relu and its gradient as C++ Ops.

lowintelligence committed:
Add LeakyRelu C++ Op and its gradient implementation.

LeakyRelu, defined as 'y = { x (x>=0) or alpha*x (x<0) }', was computed
by combined Ops 'max(x, alpha*x)' in current codes. Hence its gradient
calculation for back propagation would contain a serial of element-wise
Ops. This looks really unnecessary for such a simple op and it could be
done within just one Op with less memory accesses.

akshaym commented on tensorflow/core/kernels/relu_op.h:
```diff
@@ -131,6 +131,65 @@ void Relu6GradOp<Device, T>::OperateNoTemplate(OpKernelContext* context,
+  }
+
+ private:
+  T alpha_;
```
Per nn_ops.cc, alpha_ is always a float right?

lowintelligence replied:
Well, here I checked 'tensorflow/core/kernels/lrn_op.cc' as a reference and kept the same style with that. Since LeakyRelu C++ Op would only support {half, float, double} type, thus convert alpha type from float to T should be reasonable.

akshaym commented on tensorflow/core/ops/ops.pbtxt:
```diff
@@ -13604,6 +13604,74 @@ op {
     minimum: 1
   }
 }
+op {
```
I believe you will need to update the api_def files here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/api_def/update_api_def.sh

lowintelligence replied:
Sorry, I wrote the api_def files manually but forgot to 'git add' them. I just re-generated the files with the tool and add them to the new commit.

akshaym commented on tensorflow/python/eager/pywrap_tfe_src.cc:
```diff
@@ -1799,6 +1800,7 @@ bool OpDoesntRequireInput(const string& op_name) {
           "BiasAdd",
           "Relu",
           "Relu6",
+          "LeakyRelu",
```
The LeakyRelu gradient does seem to use the inputs? Same above.

lowintelligence replied:
My fault, 'LeakyRelu' and 'LeakyReluGrad' both use the input and don't use the output, corrected in the new commit.

akshaym commented on tensorflow/core/ops/ops.pbtxt:
```diff
@@ -13604,6 +13604,74 @@ op {
+  }
+}
+op {
+  name: "LeakykReluGrad"
```
I don't believe these changes are required (they are auto generated), but in case they are, you have a typo here - LeakykReluGrad -> LeakyReluGrad.

lowintelligence replied:
Really a typo. Thanks for figuring it out. :)

akshaym reviewed (commented):
Sorry for the delay @lowintelligence, and thanks for the contribution. 

I've just added some initial comments, and requested a more in-depth review.

lowintelligence committed:
Refine LeakyRelu codes and update APIs.

alextp commented on tensorflow/python/ops/nn_ops.py:
```diff
@@ -1601,8 +1601,7 @@ def leaky_relu(features, alpha=0.2, name=None):
     features = ops.convert_to_tensor(features, name="features")
     if features.dtype.is_integer:
       features = math_ops.to_float(features)
-    alpha = ops.convert_to_tensor(alpha, dtype=features.dtype, name="alpha")
```
This should be changed to respect our forward compatibility rules. Use the forward compatibility layer to make this change, with the date set 3 weeks in the future of the merge (so a month from now is safe): https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/compat/compat.py

lowintelligence replied:
Well, I tried to use compat.forword_compatible() on my new implementation. However, another concern (maybe over considered) is that, whether somebody would use 'alpha' as a trainable variable in his/her model? Should we try to keep compatibility for this case?

alextp commented on tensorflow/python/ops/nn_grad.py:
```diff
@@ -390,6 +390,21 @@ def _Relu6GradGrad(op, grad):
+
+
+@ops.RegisterGradient("LeakyReluGrad")
+def _LeakyReluGradGrad(op, grad):
```
Can you add a gradient checker test for these? Use https://www.tensorflow.org/api_docs/python/tf/test/compute_gradient_error

lowintelligence replied:
Well, I believe that python UTs for GradGrad was already in the original commit.

lowintelligence committed:
Refine LeakyRelu codes.

1. Add C++ gradient of gradient definition of LeakyReLu and revalant UT.
2. Using forward compatibility layer for python code changes.

lowintelligence commented on tensorflow/cc/gradients/nn_grad.cc:
```diff
@@ -143,6 +143,19 @@ Status Relu6GradHelper(const Scope& scope, const Operation& op,
+  TF_RETURN_IF_ERROR(GetNodeAttr(op.node()->attrs(), "alpha", &alpha));
+  internal::LeakyReluGrad::Attrs attrs;
+  attrs.Alpha(alpha);
+  auto dx = internal::LeakyReluGrad(scope, grad_inputs[0], op.input(0), attrs);
```
Done in new commit.

lowintelligence committed:
Code fix against ci_build error results.

lowintelligence committed:
Add XLA support for LeakyReluOp.

Code contributed by: Meng Chen <mc119496@alibaba-inc.com>

lowintelligence commented:
Due to ci_build error reports, I did some change and submit a new version here.

alextp commented on tensorflow/core/api_def/base_api/api_def_LeakyRelu.pbtxt:
```diff
@@ -0,0 +1,4 @@
+op {
+  graph_op_name: "LeakyRelu"
+  summary: "Computes rectified linear: `max(features, features * alpha)`."
+}
```
Can you add visibility:hidden here?

I think this will obviate the need for the API golden file changes since there's already tf.nn.leaky_relu and no need for tf.leaky_relu (which gets implicitly created when an op is added)

lowintelligence committed:
Avoid golden API file changing.

lowintelligence commented:
@alextp Do you have time to review the code again? Thanks.

lowintelligence committed:
Fix lint errors and typos.

lowintelligence commented:
@alextp Sorry, I have to add a commit to fix some lint errors and typos due to ci_build tests report.

lowintelligence committed:
Merge remote-tracking branch 'origin'

lowintelligence commented:
Have to merge with master due to branch conflicts.

lowintelligence committed:
Revert "Add XLA support for LeakyReluOp."

This reverts commit d2ad105d2dff3c79d8f49f5fb8ce74c38f424e74.

Since bfloat16 was not supported by LeakyRelu, but it should be
supported in XLA Ops.

lowintelligence commented:
Have to reverted XLA commit since the Op didn't provide bfloat16 support.

lowintelligence commented:
@alextp I reverted the XLA commit due to unsupported dtypes. Finally it looks that all CI build cases related with this PR was passed. Thanks for the patient of taking a review again. :)

facaiy commented:
Indeed, unrelated failures. Thanks for your work, @lowintelligence !

lowintelligence commented:
@alextp Could you please help to take a look about why "feedback/copybara" failed? I can't see detail information here.

facaiy commented:
@yifeif Hi, yifei. Do we have a plan to give contributor more details about feedback/copybara? It seems that we can't do anything but wait. cc @martinwicke

yifeif commented:
Hi @facaiy feedback/copybara is currently an experimental feature and it's a bit far from being able to show the build results with details. Please ping reviewer(s) to see if there is anything needs to be fixed from the PR. Thanks!

facaiy commented:
@yifeif Oh, I got it.  It can be quite useful when feddback/copybara is ready for showing build details. Thanks for your explanation :-)

facaiy commented:
@protoget Hi, could you lend a hand? Thanks.

alextp commented:
One internal error is error: module //tensorflow/cc:gradients_nn_grad_test does not depend on a module exporting 'tensorflow/cc/ops/nn_ops_internal.h'

I think you need to fix the build dependencies, adding something to tensorflow/cc/BUILD

lowintelligence committed:
Fix build dependencies in tensorflow/cc/BUILD.

lowintelligence commented:
Well, I've fixed the tensorflow/cc/BUILD file and delayed the forward compatibility date. Please take a check. Thanks. @alextp

lowintelligence committed:
Merge remote-tracking branch 'origin'

lowintelligence commented:
Well, I found that the commit in master branch https://github.com/tensorflow/tensorflow/commit/e2ce9787d9927e4a6574e6ac4606a47712320170 has removed the 'compat' module importing in nn_ops.py, thus many tests were failed due to this module not found. :-( I would make a merge with master and commit a new version. @alextp Thanks.

lowintelligence committed:
Re-add compat module for leaky_relu implementation.
